prov/rxm/src/rxm_cq.c

```diff
@@ -1182,17 +1182,27 @@ static void rxm_cq_read_write_error(struct rxm_ep *rxm_ep)
 		err_entry.op_context = rndv_buf->app_context;
 		err_entry.flags = ofi_tx_cq_flags(rndv_buf->pkt.hdr.op);
 		break;
-	case RXM_RNDV_ACK_SENT:
-		/* fall through */
 	case RXM_RX:
+		assert(err_entry.flags & FI_RECV);
```

Comment:
This assertion is failing only for MLX5 verbs. It passes on HFI1 verbs.
```
fi_unexpected_msg: /home/arun/scm/sfi/libfabric/..//libfabric/prov/rxm/src/rxm_cq.c:1186: rxm_cq_read_write_error: Assertion `err_entry.flags & (1ULL << 10)' failed.
```

It seems verbs work completion opcode is set to an invalid value when there is a flush error:

```
Breakpoint 2, fi_ibv_poll_cq (cq=0x66e420, wc=0x7fffffffdb90) at /home/arun/scm/sfi/libfabric/..//libfabric/prov/verbs/src/verbs_cq.c:295
295             return fi_ibv_process_wc_poll_new(cq, wc);
(gdb) p wc.status
$7 = IBV_WC_WR_FLUSH_ERR
(gdb) p wc.opcode
$8 = 32767
(gdb)
```

HFI1 verbs:
```
291             ret = ibv_poll_cq(cq->cq, 1, wc);
Value returned is $7 = 1
(gdb) p wc.status
$8 = IBV_WC_WR_FLUSH_ERR
(gdb) p wc.opcode
$9 = IBV_WC_RECV
```

Reply:
I believe WRs that complete in error only need to report the wr_id.

Reply:
It seems mlx5 is the one that's misbehaving. From Infiniband spec:

> 
> 11.4.2.1 POLL FOR COMPLETION
> ...
> Output Modifiers:
> • The Work Completion containing information relating to the completed Work Request if an entry is present on the CQ. If the status of the operation that generates the Work Completion is anything other than success, the contents of the Work Completion are undefined except as noted below. The contents of a Work Completion are:
> • The 64-bit Work Request ID set by the Consumer in the associated Work Request. This is always valid, regardless of the
> status of the operation.
 **_• The operation type specified in the completed Work Request._**
> • The valid operation types are:
> • Send (for WRs posted to the Send Queue)
> • RDMA Write (for WRs posted to the Send Queue)

May be we should just drop the assertion to accommodate mlx5.

Reply:
"the contents are undefined except as noted"  ... "WR ID ... is always valid".  I believe mlx5 is compliant.  Only the wr_id can be relied on.

Reply:
Hmm, I missed the "always valid" for just WR ID. I interpreted that the entire list of fields to be valid (those not specified in the list are undefined). I'll drop the assert then.